### PR TITLE
Do not set the compress onSend hook twice

### LINKED ISSUE #140: Corrupted reply with async prefixed routes and br encoding
## 🐛 Bug Report

The JSON response becomes corrupted when using br compression as well as prefixed async routes, This may be the unresolved cause of https://github.com/fastify/fastify-compress/issues/104

## To Reproduce

```js
// package.json
{
  "name": "node_test",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "start": "node index.js"
  },
  "dependencies": {
    "fastify": "^3.9.2",
    "fastify-compress": "^3.4.0"
  }
}
```
```js
// index.js
const server = require('fastify')();

server.register(require('fastify-compress'), {
    threshold: 0,
    // the issue disappears by excluding 'br' from the encodings
    // encodings: ['gzip', 'deflate', 'identity']
});

server.register(require('./routes'));

server.listen(8080);
```
```js
// routes/index.js
module.exports = async (fastify, _) => {
    fastify.register(require('./test'), {
        prefix: '/test', // if this prefix is removed, the issue disappears
    });
}
```
```js
// routes/test.js
module.exports = async (fastify, opts) => {
    fastify.get('/', opts, async (req, _) => {
        return { hi: true };
    });
}
```
If a GET request is made to `http://localhost:8080/test` the response body is:
```js
�{"hi":true}
```
Note the expected behavior is produced if the prefix in routes/index.js is removed, or if 'br' encoding is explicitly removed when registering fastify-compress

## Expected behavior

```js
{"hi":true}
```

## Your Environment

- *node version*: v14.15.1
- *fastify version*: >=3.9.2
- *os*: Windows 10
- *GET request was made with the chrome fetch API & postman and results were the same*

### PR DESCRIPTION
**Pull request by mcollina**
Fixes https://github.com/fastify/fastify-compress/issues/140

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -179,6 +181,16 @@ function processDecompressParams (opts) {
 }
 
 function buildRouteCompress (fastify, params, routeOptions, decorateOnly) {
+  // This methods works by altering the routeOptions, it has side effects.
```
Comment (jsumners) [inline index.js:184]: I don't understand how this is getting added twice and why the side effect is necessary. The test is only registering the plugin once. Can you provide detail?

File: test/test-routes-compress.js [hunk 2]
```diff
@@ -255,3 +255,32 @@ test('should throw an error on invalid compression setting', t => {
...
+
+test('avoid double onSend', t => {
+  t.plan(2)
+
+  const server = Fastify()
+
+  server.register(compressPlugin, {
+    threshold: 0
+    // the issue disappears by excluding 'br' from the encodings
+    // encodings: ['gzip', 'deflate', 'identity']
```
Comment (Eomm) [inline test/test-routes-compress.js:267]:
```suggestion
```

File: index.js (hunk 1 above)
Comment (Eomm) [reply index.js:184]: It should be caused by the `ignoreTrailingSlash` that register the routes `/hello` and `/hello/`

Comment (zekth) [reply index.js:184]: i Confirm : 
```
/test
/test/
```


```js
    function afterRouteAdded (path, notHandledErr, done) {
      const url = prefix + path
      console.log(url)
```

**Review by RafaelGSS — APPROVED** [review]
LGTM.

File: index.js (hunk 1 above)
Comment (jsumners) [reply index.js:184]: Excellent. That reduces this to the actual issue.

**Review by jsumners — APPROVED** [review]
LGTM

If there's a way to do this without mutating the options object, though, I'd prefer to see that.

**Review by delvedor — APPROVED** [review]
LGTM

File: index.js (hunk 1 above)
Comment (zekth) [reply index.js:184]: This might be good to be documented in https://www.fastify.io/docs/latest/Hooks/#onroute don't you think? Like a `Note`.

**Review by zekth — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
> If there's a way to do this without mutating the options object, though, I'd prefer to see that.

That will require a semver-major change in Fastify - we should probably do it.

**Review by Ethan-Arrowood — APPROVED** [review]

**Pull request merged**